ansible_rulebook/engine.py

```diff
@@ -156,16 +156,32 @@ async def start_source(
             raise Exception("Entrypoint is not a coroutine function.")
 
         await entrypoint(fqueue, args)
+        shutdown_msg = (
+            f"Source {source.source_name} initiated shutdown at "
+            f"{str(datetime.now())}"
+        )
 
     except KeyboardInterrupt:
+        shutdown_msg = (
+            f"Source {source.source_name} keyboard interrupt, "
+            f"initiated shutdown at {str(datetime.now())}"
+        )
         pass
     except asyncio.CancelledError:
-        logger.info("Task cancelled")
-    except BaseException:
+        shutdown_msg = (
+            f"Source {source.source_name} task cancelled, "
+            "initiated shutdown at {str(datetime.now())}"
+        )
+        logger.info("Task cancelled " + shutdown_msg)
+    except BaseException as e:
         logger.exception("Source error")
+        shutdown_msg = (
+            f"Shutting down source : {source.source_name} error : {str(e)}"
```

Comment:
```suggestion
            f"Shutting down source: {source.source_name} error: {str(e)}"
```

Reply:
No need to cast str in f-strings, it's implicit:
```f"Shutting down source: {source.source_name} error: {e}"```